data/basal.go

```
@@ -0,0 +1,166 @@
+
+	levemirInsulin, lantusInsulin = "levemir", "lantus"
+
+	rateValidationLowerLimit     = 0.0
+	durationValidationLowerLimit = 0
+	valueValidationLowerLimit    = 0
+	validationGreaterThanInt     = "Must be greater than %d"
+)
+
+var (
+	allowedDeliveryTypes = map[string]string{injectedDelivery: injectedDelivery, scheduledDelivery: scheduledDelivery, suspendDelivery: suspendDelivery, tempDelivery: tempDelivery}
+	allowedInsulins      = map[string]string{levemirInsulin: levemirInsulin, lantusInsulin: lantusInsulin}
+
+	basalFailureReasons = validate.ErrorReasons{
+		deliveryTypeTag: fmt.Sprintf("Must be one of %s,%s,%s,%s", injectedDelivery, scheduledDelivery, suspendDelivery, tempDelivery),
+		rateTag:         fmt.Sprintf("Must be greater than %f", rateValidationLowerLimit),
+		durationTag:     fmt.Sprintf(validationGreaterThanInt, durationValidationLowerLimit),
+		valueTag:        fmt.Sprintf(validationGreaterThanInt, valueValidationLowerLimit),
+		insulinTag:      fmt.Sprintf("Must be one of %s,%s", levemirInsulin, lantusInsulin),
+	}
+)
+
+//BuildBasal will build a Basal record
+func BuildBasal(obj map[string]interface{}) (*Basal, *Error) {
+
+	base, errs := BuildBase(obj)
+	cast := NewCaster(errs)
+
+	basal := &Basal{
+		DeliveryType: cast.ToString(deliveryTypeField, obj[deliveryTypeField]),
+		ScheduleName: cast.ToString(scheduleNameField, obj[scheduleNameField]),
+		Base:         base,
+	}
+
+	if obj[rateField] != nil {
+		basal.Rate = cast.ToFloat64(rateField, obj[rateField])
+	}
+	if obj[durationField] != nil {
+		basal.Duration = cast.ToInt(durationField, obj[durationField])
+	}
+	if obj[insulinField] != nil {
+		basal.Insulin = cast.ToString(insulinField, obj[insulinField])
+	}
+
+	if validationErrors := validator.Struct(basal); len(validationErrors) > 0 {
+		errs.AppendError(validationErrors.GetError(basalFailureReasons))
+	}
+
+	if errs.IsEmpty() {
+		return basal, nil
```

My general policy is to never return the object if you are returning and error.  The problem is that if the caller ignores the errors they still get a valid object back (even though it isn't really valid).  So, I'd like something like:

```
if !errs.IsEmpty() {
  return nil, errs
}
return basal, nil
```